File: coding_skill_cpp/07_observer_pattern.cpp
/* C++ Observer - vector<function> */
#include <iostream>
#include <vector>
#include <functional>
#include <algorithm>
using namespace std;

template<typename T>
class Subject {
    vector<function<void(const T&)>> observers;
    T data;
    
public:
    void attach(function<void(const T&)> obs) {
        observers.push_back(obs);
    }
    
    void setData(const T& newData) {
        data = newData;
        notify();
    }
    
    void notify() {
        cout << "[Subject] 통지: " << observers.size() << "명" << endl;
        for (auto& obs : observers) {
            obs(data);
        }
    }
};

int main() {
    cout << "\n=== C++ Observer Pattern ===" << endl;
    
    Subject<int> subject;
    
    subject.attach([](const int& val) {
        cout << "  → Observer 1: " << val << endl;
    });
    
    subject.attach([](const int& val) {
        cout << "  → Observer 2: " << val * 2 << endl;
    });
    
    subject.setData(10);
    subject.setData(20);
    
    return 0;
}

File: coding_skill_py/07_observer_pattern.py
"""Python Observer - weakref와 property"""
from typing import List, Callable, Generic, TypeVar
import weakref

T = TypeVar('T')

class Subject(Generic[T]):
    def __init__(self):
        self._observers: List[Callable] = []
        self._data: T = None
    
    def attach(self, observer: Callable):
        self._observers.append(observer)
    
    def detach(self, observer: Callable):
        self._observers.remove(observer)
    
    @property
    def data(self) -> T:
        return self._data
    
    @data.setter
    def data(self, value: T):
        self._data = value
        self._notify()
    
    def _notify(self):
        print(f"[Subject] 통지: {len(self._observers)}명")
        for obs in self._observers:
            obs(self._data)

def main():
    print("\n=== Python Observer Pattern ===")
    
    subject = Subject[int]()
    
    subject.attach(lambda val: print(f"  → Observer 1: {val}"))
    subject.attach(lambda val: print(f"  → Observer 2: {val * 2}"))
    
    subject.data = 10
    subject.data = 20

if __name__ == "__main__":
    main()
